File: sea-shell-service-immutable/src/main/java/org/jesperancinha/shell/webflux/immutable/repository/ShellImmutableRepository.java
package org.jesperancinha.shell.webflux.immutable.repository;

import org.jesperancinha.shell.client.shells.Shell;
import org.jesperancinha.shell.client.shells.ShellsClient;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.publisher.ParallelFlux;

import static reactor.core.scheduler.Schedulers.elastic;
import static reactor.core.scheduler.Schedulers.single;

@Repository
@ConditionalOnBean(ShellsClient.class)
public class ShellImmutableRepository {

    private final ShellsClient shellsClient;

    @Value("${sea.shell.parallelism:20}")
    private Integer parallelism;

    public ShellImmutableRepository(ShellsClient shellsClient) {
        this.shellsClient = shellsClient;
    }

    public Mono<Shell> findSeaShellById(final Long id) {
        return Mono.fromCallable(() -> shellsClient.getShell(id)).subscribeOn(elastic());
    }

    public ParallelFlux<Shell> findAllSeaShells() {
        return findAllShellIds()
                .parallel(parallelism)
                .runOn(elastic())
                .map(shellsClient::getShell)
                .runOn(single());
    }

    public Flux<Long> findAllShellIds() {
        return Mono.fromCallable(shellsClient::getAllShellIds)
                .flux().flatMap(Flux::fromIterable);
    }
}

File: sea-shell-service-immutable/src/main/kotlin/org/jesperancinha/shell/webflux/immutable/repository/ShellImmutableRepository.kt
package org.jesperancinha.shell.webflux.immutable.repository

import org.jesperancinha.shell.client.shells.Shell
import org.jesperancinha.shell.client.shells.ShellsClient
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
import org.springframework.stereotype.Repository
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.publisher.ParallelFlux
import reactor.core.scheduler.Schedulers

@Repository
@ConditionalOnBean(ShellsClient::class)
class ShellImmutableRepository(private val shellsClient: ShellsClient) {
    @Value("\${sea.shell.parallelism:20}")
    private val parallelism: Int? = null
    fun findSeaShellById(id: Long?): Mono<Shell?> {
        return Mono.fromCallable { shellsClient.getShell(id) }.subscribeOn(Schedulers.boundedElastic())
    }

    fun findAllSeaShells(): ParallelFlux<Shell> {
        return findAllShellIds()
            .parallel(parallelism!!)
            .runOn(Schedulers.boundedElastic())
            .map { id: Long? -> shellsClient.getShell(id) }
            .runOn(Schedulers.single())
    }

    fun findAllShellIds(): Flux<Long?> {
        return Mono.fromCallable { shellsClient.allShellIds }
            .flux().flatMap { it: List<Long>? -> Flux.fromIterable(it) }
    }
}